scheduler/replica_scheduler.go

```diff
@@ -203,48 +214,77 @@ func (rcs *ReplicaScheduler) getDiskCandidates(nodeInfo map[string]*longhorn.Nod
-			return diskCandidates
+		verboseErrors.AppendErrors(errors)
+		diskCandidates, errors = getDiskCandidatesFromNodes(filterNodesWithLessThanTwoReplicas(nodesWithEvictingReplicas))
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
+		verboseErrors.AppendErrors(errors)
 	case !zoneSoftAntiAffinity && nodeSoftAntiAffinity:
-		if diskCandidates := getDiskCandidatesFromNodes(unusedNodesInNewZones); len(diskCandidates) > 0 {
-			return diskCandidates
+		diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInNewZones)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
-		if diskCandidates := getDiskCandidatesFromNodes(nodesInUnusedZones); len(diskCandidates) > 0 {
-			return diskCandidates
+		verboseErrors.AppendErrors(errors)
+		diskCandidates, errors = getDiskCandidatesFromNodes(nodesInUnusedZones)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
-		if diskCandidates := getDiskCandidatesFromNodes(nodesWithEvictingReplicas); len(diskCandidates) > 0 {
-			return diskCandidates
+		verboseErrors.AppendErrors(errors)
+		diskCandidates, errors = getDiskCandidatesFromNodes(nodesWithEvictingReplicas)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
+		verboseErrors.AppendErrors(errors)
 	case zoneSoftAntiAffinity && nodeSoftAntiAffinity:
-		if diskCandidates := getDiskCandidatesFromNodes(unusedNodesInNewZones); len(diskCandidates) > 0 {
-			return diskCandidates
+		diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInNewZones)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
-		if diskCandidates := getDiskCandidatesFromNodes(unusedNodes); len(diskCandidates) > 0 {
-			return diskCandidates
+		verboseErrors.AppendErrors(errors)
+		diskCandidates, errors = getDiskCandidatesFromNodes(unusedNodes)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
-		if diskCandidates := getDiskCandidatesFromNodes(usedNodes); len(diskCandidates) > 0 {
-			return diskCandidates
+		verboseErrors.AppendErrors(errors)
+		diskCandidates, errors = getDiskCandidatesFromNodes(usedNodes)
+		if len(diskCandidates) > 0 {
+			return diskCandidates, nil
 		}
+		verboseErrors.AppendErrors(errors)
 	}
-	return map[string]*Disk{}
+	return map[string]*Disk{}, verboseErrors
 }
 
-func (rcs *ReplicaScheduler) filterNodeDisksForReplica(node *longhorn.Node, disks map[string]struct{}, replicas map[string]*longhorn.Replica, volume *longhorn.Volume, requireSchedulingCheck bool) map[string]*Disk {
-	preferredDisk := map[string]*Disk{}
+func (rcs *ReplicaScheduler) filterNodeDisksForReplica(node *longhorn.Node, disks map[string]struct{}, replicas map[string]*longhorn.Replica, volume *longhorn.Volume, requireSchedulingCheck bool) (preferredDisks map[string]*Disk, verboseErrors util.VerboseErrors) {
+	verboseErrors = util.NewVerboseErrors()
+	preferredDisks = map[string]*Disk{}
+
+	if len(disks) == 0 {
+		verboseErrors[longhorn.SchedulerErrorNoDisAvailable] = struct{}{}
```

Comment:
The same, if possible, add a formatted error w/ node name, so it's better to know which node has no disks available? 

The same as below errors.

Reply:
If there are tens or hundreds of disks, the annotation become very lengthy. IMO, the final result is that the replica can be scheduled in each disk and on each node. Current annotation value gives enough clues and can helps the administrator to understand what happen in the system.
For more information, user can the details in longhorn logs.

Reply:
Agreed that the concern is fair. Let's keep it as is, but just ensure there will be no duplicated errors in the aggregated error.